pkg/jobworker/worker.go

```diff
@@ -0,0 +1,219 @@
+	}
+	j.cmd.Stdout = f
+	j.cmd.Stderr = f
+	// Run the command as a given user as not to escalate privilege, since the executing user must also manage cgroups
+	if WORKER_UID != -1 && WORKER_GUID != -1 {
+		j.cmd.SysProcAttr.Credential = &syscall.Credential{Uid: uint32(WORKER_UID), Gid: uint32(WORKER_GUID)}
+	}
+	j.cmd.SysProcAttr.Setpgid = true
+	// Start the job
+	err = j.cmd.Start()
+	if err != nil {
+		return nil, fmt.Errorf("failed to start job's exec.Cmd: %w", err)
+	}
+	// Assign the process group ID to the job so that we have a reference to signal child processes in Stop if the command quits
+	j.pgid, err = syscall.Getpgid(j.cmd.Process.Pid)
+	if err != nil {
+		fmt.Printf("error getting pgid: %v", err)
+		return nil, err
+	}
+	// Run go routine to handle the blocking call exec.Cmd.Wait() and update the running flag to indicate the job has complete
+	go func(runningJob *Job, logFile *os.File) {
+		runningJob.cmd.Wait()
+		runningJob.setRunning(false)
+		runningJob.done <- true
+		logFile.Close()
+	}(j, f)
+	return j, nil
+}
+
+// Stop request a job's termination using SIGTERM and deletes it's cgroup. If the SIGTERM is ignored, we send
+// a SIGKILL after STOP_GRACE_PERIOD.
+func (job *Job) Stop() error {
+	// Regardless of signalling errors, ensure we clean up the job's log file and cgroup
+	defer func() {
+		os.Remove(logPath(job.ID))
+		job.con.DeleteGroup(job.ID)
+	}()
+	// Send SIGTERM to process group
+	if err := syscall.Kill(-job.pgid, syscall.SIGTERM); err != nil {
+		return err
+	}
+	// Set up timeout
+	ctx, cancel := context.WithTimeout(context.Background(), STOP_GRACE_PERIOD)
+	defer cancel()
+	// If job is running potentially wait on either done channel or SIGKILL after timeout
+	if job.isRunning() {
+		select {
+		case <-job.done:
+			break
+		case <-ctx.Done():
+			if err := syscall.Kill(-job.pgid, syscall.SIGKILL); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// Status generates a JobStatus with information from the job and it's underlying os.Process & os.ProcessState
+func (job *Job) Status() JobStatus {
+	// Get PID and possible exit code from Process and ProcessState, assume running if ProcessState is nil
+	pid := 0
+	if job.cmd.Process != nil {
+		pid = job.cmd.Process.Pid
+	}
+	// Check if running flag has been set after blocking Wait call on job.cmd
+	running := job.isRunning()
+	exitCode := 0
+	if !running {
+		exitCode = job.cmd.ProcessState.ExitCode()
+	}
+	return JobStatus{
+		ID:       job.ID,
+		PID:      pid,
+		Running:  running,
+		ExitCode: exitCode,
+	}
+}
+
+// Output returns a wrapped io.ReadCloser that "tails" the job's log file
+// If follow=true, the Read will block and poll for updates to the file. It is then the responsibility
+// of the caller to ensure .Close is called to prevent the Read blocking indefinitely.
+// If follow=false, upon Read'ing the entire file an io.EOF will be returned
+func (job *Job) Output(follow bool) (reader io.ReadCloser, err error) {
```

Comment:
I have a personal phobia of `bool` parameters, as they're usually completely uninformative at the call site. Maybe a 2-state enum?

```golang
type OutputMode int
const (
  OutputModeNoFollow OutputMode = 0 // feel free to supply a better name
  OutputModeFollow   OutputMode = 1
)
```

```suggestion
func (job *Job) Output(mode OutputMode) (reader io.ReadCloser, err error) {
```


which makes the call site go from 

```golang
job.Output(true)
```

to 

```golang
job.Output(OutputModeFollow)
```

Reply:
Yes I like the look of that